sdl-kafka/src/main/scala/io/smartdatalake/workflow/dataobject/KafkaTopicDataObject.scala

```diff
@@ -347,21 +349,21 @@ case class KafkaTopicDataObject(override val id: DataObjectId,
     dfs.reduce(_ union _)
   }
 
-  private def convertToWriteDataFrame(df: DataFrame): DataFrame = {
+  private def encodeKeyValue(df: DataFrame): DataFrame = {
     require(df.columns.toSet == Set("key","value"), s"($id) Expects columns key, value in DataFrame for writing to Kafka. Given: ${df.columns.mkString(", ")}")
-    keySchema.foreach(schema => validateSchema(schema, SparkSchema(df.schema("key").dataType.asInstanceOf[StructType]), "read (keySchema)"))
-    valueSchema.foreach(schema => validateSchema(schema, SparkSchema(df.schema("value").dataType.asInstanceOf[StructType]), "read (valueSchema)"))
+    keySchema.foreach(schema => validateSchema(schema, SparkSchema(df.schema("key").dataType.asInstanceOf[StructType]), "write (keySchema)"))
+    valueSchema.foreach(schema => validateSchema(schema, SparkSchema(df.schema("value").dataType.asInstanceOf[StructType]), "write (valueSchema)"))
     df.select(
       convertToKafka(keyType, col("key"), SubjectType.key, keySchema).as("key"),
-      convertToKafka(valueType, col("value"), SubjectType.value, valueSchema).as("value")
+      when(col("value").isNotNull, convertToKafka(valueType, col("value"), SubjectType.value, valueSchema)).as("value")
```

Comment:
When decoding the value in *decodeKeyValue*, we exclude values that are NULL **and** the ones that have length == 0. Here, we only convert nonNull values to Kafka. Why the difference?

Reply:
Good question.
In decodeKeyValue a Binary column has to be decoded. If its value is an empty binary array, decoding probably fails.
In encodeKeyValue it is a column of arbitrary type. Checking for length is possible for Array, String and Binary, but is not defined for Struct or Numeric types.
To have the same definition we could check length=0 if it is a binary type. I have already adapted the code.